File: erlang/lectures/parallel/src/fib.erl
-module(fib).

-compile(export_all).



fib(0) -> 1;
fib(1) -> 1;
fib(N) ->  
    F1 = fib(N-1),
    F2 = fib(N-2),
    F1 + F2.








fip(0) -> 1;
fip(1) -> 1;
fip(N)  -> 
    Ref = make_ref(),
    parallel(fun() -> fip(N-1) end, Ref),
    parallel(fun() -> fip(N-2) end, Ref),
    F1 = collect(Ref),
    F2 = collect(Ref),
    F1 + F2.









fix(0,_) -> 1;
fix(1,_) -> 1;
fix(N,M) when N > M->
    Ref = make_ref(),
    parallel(fun() -> fix(N-1, M) end, Ref),
    parallel(fun() -> fix(N-2, M) end, Ref),
    F1 = collect(Ref),
    F2 = collect(Ref),
    F1 + F2;
fix(N,_)  ->  fib(N).

parallel(F, Ref) ->
    Self = self(),
    spawn(fun() -> Res = F(), Self ! {ok, Ref, Res} end).

collect(Ref) ->
    receive
	{ok, Ref, Res} ->
	    Res
    end.







naive() ->	
    {Fib, _} = timer:tc(fun() -> fib(30) end),
    {Fip, _} = timer:tc(fun() -> fip(30) end),
    io:format("fib(30)  sequential: ~w ms, naive parallel ~w ms~n", [Fib div 1000, Fip div 1000]).


better() ->	
    {Fib, _} = timer:tc(fun() -> fib(40) end),
    io:format("fib(40)  sequential: ~w ms~n", [Fib div 1000]),
    bench(40,38),
    bench(40,36),
    bench(40,34),
    bench(40,32),
    bench(40,30),
    bench(40,28),
    bench(40,26),
    bench(40,24),
    bench(40,22),
    bench(40,20).

bench(N, M) ->
    {T, _} = timer:tc(fun() -> fix(N, M) end),
    io:format("fib(~w,~w) : ~w ms~n", [N, M, T div 1000]).

    

    
    
		     
			
		  
     
    

File: lectures/parallel/src/fib.ex
defmodule Fib do

  def sequential(n) do
    {t,f} = :timer.tc(fn() -> seq(n) end)
    IO.puts("sequential fib(#{n}) = #{f}  in #{trunc(t/1000)} ms")
    :ok
  end

  def parallel(n) do
    {t,f} = :timer.tc(fn() -> par(n) end)
    IO.puts("parallel fib(#{n}) = #{f}  in #{trunc(t/1000)} ms")
    :ok
  end

  def combined(n, k) do
    {t,f} = :timer.tc(fn() -> comb(n, k) end)
    IO.puts("combined fib(#{n}) = #{f}  in #{trunc(t/1000)} ms")
    :ok
  end  

  def comb(0,_) do 0 end
  def comb(1,_) do 1 end
  def comb(n, k) when n > k do
    r1 = paral(fn() -> comb(n-1, k) end)
    r2 = paral(fn() -> comb(n-2, k) end)    
    f1 = collect(r1)
    f2 = collect(r2)
    f1 + f2
  end
  def comb(n, _) do
    seq(n)
  end


  
  def seq(0) do 0 end
  def seq(1) do 1 end
  def seq(n) do
    seq(n-1) + seq(n-2)
  end

  def par(0) do 0 end
  def par(1) do 1 end
  def par(n) do
    r1 = paral(fn() -> par(n-1) end)
    r2 = paral(fn() -> par(n-2) end)    
    f1 = collect(r1)
    f2 = collect(r2)
    f1 + f2
  end    

  def paral(f) do
    ref = make_ref()
    me = self()
    spawn(fn() -> send(me, {:par, ref, f.()}) end)
    ref
  end
      
  def collect(ref) do
    receive do
      {:par, ^ref, res} -> res
    end
  end


  def bench(n, k) do
    Enum.each(n..k, fn(s) ->
      {c,_} = :timer.tc(fn() -> comb(n, s) end)
      IO.puts(" k = #{s} in #{trunc(c/1000)} ms")
    end)
    :ok
  end


  def graph(p) do
    {:ok, fd} = File.open("fib.dat", [:write])
    IO.write(fd, "# fib(40,30), exeution time in ms\n")
    IO.write(fd, "#proc\tmin\tq2\tq4\tmax\n")    
    Enum.each(1..p, fn(p) -> graph(p, fd) end)
    File.close(fd)
  end

  def graph(p, fd) do
    :erlang.system_flag(:schedulers_online, p)
    res = Enum.map(1..100, fn(_) -> {t,_} = :timer.tc(fn() -> comb(40,30) end); t end)
    res = Enum.sort(res)
    min = trunc(:lists.nth(1, res)/1000)
    q2 = trunc(:lists.nth(25, res)/1000)
    q4 = trunc(:lists.nth(75, res)/1000)
    max = trunc(:lists.nth(100,res)/1000)
    IO.write(fd, "#{p}\t#{min}\t#{q2}\t#{q4}\t#{max}\n")
  end

  
end
